pkg/search/controllers_test: reduce the probability of flaky test cases by waiting for cache to sync

mohamedawnallah opened the pull request:
**What type of PR is this?**

/kind flake

<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:

This PR reduce the probability of flaky test cases by waiting for cache to sync. It seems fake client has some flaky issues with watch implementation that is different from real implementation as seen in https://github.com/kubernetes/kubernetes/issues/87472#issuecomment-577472496?

**Which issue(s) this PR fixes**:
Fixes #5926

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
NONE
```

XiShanYongYe-Chang commented on pkg/search/controllers_test.go:
```diff
@@ -577,11 +584,12 @@ func TestDeleteResourceRegistryEventHandler(t *testing.T) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to create new controller, got: %v", err)
 	}
+	newController.InformerManager = genericmanager.NewMultiClusterInformerManager(stopCh)
```
Is this modification unnecessary? newController.InformerManager has been initialized.

mohamedawnallah replied:
The new multicluster informer manager global instance is created using the singleton pattern (`once`), which ensures it is instantiated only once. As a result, test cases depend on this single instance, leading to unintended dependencies between the tests. What do you think?

https://github.com/karmada-io/karmada/blob/8d89d77110cf378d09b5f43dcb2e09bb393d61f1/pkg/util/fedinformer/genericmanager/multi-cluster-manager.go#L39-L44

XiShanYongYe-Chang replied:
Yes, you're right, it's right not to share data between each test case.

XiShanYongYe-Chang commented:
There is indeed such a setting in the controller startup, and I think this should work:

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/search/controller.go#L136

zhzhuang-zju commented:
I executed `go test -v pkg/search/controller.go pkg/search/controllers_test.go --count=2000` twice in my environment, and it finally failed.

XiShanYongYe-Chang commented:
In terms of implementation, the WaitForCacheSync method should be called. There may be other problems with the ut failure. I feel that I can continue to push this pr and explore other problems.

Hi @mohamedawnallah can you revise it according to the comments?

mohamedawnallah commented:
> I executed go test -v pkg/search/controller.go pkg/search/controllers_test.go --count=2000 twice in my environment, and it finally failed.

Thanks, @zhzhuang-zju, for testing this 2000 times, twice! You're right. I’ve also tried this, and as mentioned in the PR description, the proposed changes reduce the likelihood of flaky test cases. Before, the tests were failing more frequently. It would be interesting to gather more empirical observations from the GitHub CI test failures as @XiShanYongYe-Chang  pointed, if any. What are your thoughts?

mohamedawnallah commented:
> In terms of implementation, the WaitForCacheSync method should be called. There may be other problems with the ut failure. I feel that I can continue to push this pr and explore other problems.

Currently, the `Start` method on the controller is busy-waiting for a signal from `stopCh`, and during this process, it caches the next element every second. After testing this, I encountered some timing issues. We might need to introduce static sleep in the verification code to allow time for the busy-waiting goroutine to execute the caching. This could reduce control over the verification code in the test cases. For instance, when a cluster is upserted, we would want to cache it immediately after.

What do you think, @XiShanYongYe-Chang?

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/search/controller.go#L131-L145

mohamedawnallah commented:
Apologies for the delayed response to the PR feedback. I've been quite busy recently, but I’ve now added comments addressing all the feedback. I’d love to hear your thoughts on them. 🙏

XiShanYongYe-Chang commented:
> We might need to introduce static sleep in the verification code to allow time for the busy-waiting goroutine to execute the caching.

I don't quite understand why we need to add static time to wait. Events are blocked waiting, and as long as events can be fetched from the queue, processing is executed. So I understand that the changes now are sufficient.

mohamedawnallah commented:
> I don't quite understand why we need to add static time to wait. Events are blocked waiting, and as long as events can be fetched from the queue, processing is executed. So I understand that the changes now are sufficient.

Sorry for not providing enough context! 🙏 I was referring to what happens if we use `controller.Start(test.stopCh)`.

XiShanYongYe-Chang commented:
Oh! Thanks @mohamedawnallah 

> Apologies for the delayed response to the PR feedback.

It's okay. As long as you have time, you can always contribute.

zhzhuang-zju commented:
thanks
/lgtm
